### PR TITLE
[lld][WebAssembly] Always search *.so for -Bdynamic

### PR DESCRIPTION
**Pull request by yamt**
Search *.so libraries regardless of -pie to make it a bit more straightforward to build non-pie dynamic-linked executables.

Flip the default to -Bstatic (unless -pie or -shared is specified) as I think it's what most users expect for the default as of today.
The assumption here is that, because dynamic-linking is not widely used for WebAssembly, the most users do not specify -Bdynamic or -Bstatic, expecting static link.
Although the recent wasi-sdk ships *.so files, there are not many wasm runtimes providing the support of dynamic-linking. (only emscripten and toywasm as far as i know.)

### CODE REVIEW

**Review by sbc100 — COMMENTED** [review]
Seems like a good change.

How about we keep `-Bdynamic` as the default when building with `-shared` and/or `-pie` ?  It seems like in those modes we want to consume shared libraries by default.

Another alternative is that we always prefer shared libraries unless `-static` is passed? then it would be up to the clang driver (or users) to specify `-static` in the presence of both `.so` and `.a` files.

Ideally we would have some flag

**Comment by yamt** [comment]
> Seems like a good change.
> 
> How about we keep `-Bdynamic` as the default when building with `-shared` and/or `-pie` ? It seems like in those modes we want to consume shared libraries by default.

it sounds reasonable to me.

> Another alternative is that we always prefer shared libraries unless `-static` is passed? then it would be up to the clang driver (or users) to specify `-static` in the presence of both `.so` and `.a` files.

as the latest wasi-sdk ships *.so files, i guess it surprises many users.

> Ideally we would have some flag

what do you mean?

**Comment by MaskRay** [comment]
In ELF linkers, -Bdynamic is the default. Clang Driver `-static` passes `-static` to the linker.

**Comment by sbc100** [comment]
> In ELF linkers, -Bdynamic is the default. Clang Driver `-static` passes `-static` to the linker.

The problem is that dynamic linking under wasm is still experimental so we probably want to keep the default as static for now.

IIUC we have several options for how to achieve "opt-in" dynamic linking:

1. Make `-Bstatic` the default.   The downside of this is that it will be annoying if we ever want to change it in the future.
2. Don't skip `.so` files in the normal wasi-sdk, but only in an experimental one.  Users of the experimental sdk would have to know that they need to pass `-static` to get the current behaviour.
3. Have the clang driver enforce the "static-by-default" by having it inject `-static` for wasm/wasi targets.   This has the same downside as (1) except direct callers of wasm-ld won't be effected by any change in defaults.

The simplest option for now seem like (1).

File: lld/wasm/Config.h [hunk 1]
```diff
@@ -72,7 +72,7 @@ struct Configuration {
   bool stripAll;
   bool stripDebug;
   bool stackFirst;
-  bool isStatic = false;
+  bool isStatic = true;
```
Comment (sbc100) [inline lld/wasm/Config.h:75]: Perhaps a comment:  "Because dyamanic linking under Wasm is still experimental we default to static linking"

**Review by sbc100 — COMMENTED** [review]
I wonder if there is a test we can write for this?

I'm actually not sure what would happen if you did `-Bdynamic` without `-shared` or `-pie` .. I don't think that is something that is currently supported.  Have you tried it?

**Comment by yamt** [comment]
> I'm actually not sure what would happen if you did -Bdynamic without -shared or -pie .. I don't think that is something that is currently supported. Have you tried it?

yes. it has been working for toywasm at least.
https://github.com/yamt/garbage/blob/1ddf7388fc91c31beed54101685e5d90573b43ef/c/shlib/wasi.sh#L70-L91

File: lld/wasm/Config.h (hunk 1 above)
Comment (yamt) [reply lld/wasm/Config.h:77]: done

**Review by sbc100 — APPROVED** [review]

File: lld/wasm/Driver.cpp [hunk 2]
```diff
@@ -556,6 +554,11 @@ static void readConfigs(opt::InputArgList &args) {
   config->zStackSize =
       args::getZOptionValue(args, OPT_z, "stack-size", WasmPageSize);
 
+  // -Bdynamic by default if -pie or -shared is specified.
+  if (config->pie || config->shared) {
+    config->isStatic = false;
+  }
```
Comment (MaskRay) [inline lld/wasm/Driver.cpp:560]: The convention removes braces in this single line statement case.

**Review by MaskRay — APPROVED** [review]

File: lld/wasm/Driver.cpp (hunk 2 above)
Comment (yamt) [reply lld/wasm/Driver.cpp]: fixed. thank you.

**Comment by dschuff** [comment]
Yes, it looks ready. Do you just need someone to land it? Can you make sure the commit message is up-to-date in the top comment?

**Comment by yamt** [comment]
> Yes, it looks ready. Do you just need someone to land it?

i think so, yes.

> Can you make sure the commit message is up-to-date in the top comment?

i updated the description a bit.

**Pull request merged**